File: backend-go/projects/gateway/internal/gatewaydispatch/candidatefilter.go
package gatewaydispatch

import (
	"context"

	"github.com/huanminabc/juhe-ai/backend-go-gateway/internal/gatewaypreauth"
	"github.com/huanminabc/juhe-ai/backend-go-gateway/internal/gatewayrouting"
)

// filterOpenAIGatewayRequestCandidateAccounts, migrated from
// dispatch/candidate-filter.ts.

// CandidateFilterOutput mirrors RequestCandidateFilterResult.
type CandidateFilterOutput struct {
	// Outcome is 'accounts' | 'fallback' | 'completed'.
	Outcome       string
	Accounts      []AccountCandidate
	ModelPriority *gatewayrouting.GatewayAccountModelPriority
	// Fallback variant
	Reason  string
	Context any
}

// CandidateFilterArgs mirrors the Node input object.
type CandidateFilterArgs struct {
	Req                        *gatewaypreauth.GatewayRequest
	AuditCapture               AuditCapture
	UsageContext               gatewaypreauth.GatewayFailureUsageContext
	StartedAt                  int64
	RawCandidateAccounts       []AccountCandidate
	ClientStrategy             gatewaypreauth.ClientStrategyContext
	SystemAccountID            string
	APIKeyID                   string
	GroupID                    string
	ClientIP                   string
	Endpoint                   string
	BypassModelFilter          bool
	RequestModelOverride       string
	RouteCoordinator           gatewayrouting.GatewayRouteCoordinatorOwner
	RecoverUnavailableCandidateAccounts func(ctx context.Context) ([]AccountCandidate, error)
	LoadModelAwareCandidateAccounts     func(ctx context.Context, requestedModel, sourceEndpointFamily string) ([]AccountCandidate, error)
}

// requestCapabilityMismatchMessage mirrors requestCapabilityMismatchMessage.
func requestCapabilityMismatchMessage(reason string) string {
	if reason == "anthropic_native_group_openai_compatible_request" {
		return "当前 API Key 绑定的是 Anthropic 原生分组，不兼容 Codex / OpenAI 请求路径；请改用 Anthropic /v1/messages 客户端，或绑定支持 OpenAI Responses / Chat Completions 的分组"
	}
	return "当前分组无账户支持请求路径或客户端协议"
}

// shouldReloadModelAwareCandidates mirrors shouldReloadModelAwareCandidates.
func shouldReloadModelAwareCandidates(requestedModel string, filter ModelFilterResult, loader func(ctx context.Context, requestedModel, sourceEndpointFamily string) ([]AccountCandidate, error)) bool {
	return requestedModel != "" && loader != nil &&
		filter.DirectMatchedCount == 0 && filter.MappingMatchedCount == 0
}

// FilterOpenAIGatewayRequestCandidateAccounts mirrors
// filterOpenAIGatewayRequestCandidateAccounts.
func (p *CandidatePipeline) FilterOpenAIGatewayRequestCandidateAccounts(ctx context.Context, input CandidateFilterArgs) (CandidateFilterOutput, error) {
	requestedModel := trimString(input.RequestModelOverride)
	if requestedModel == "" {
		if model, ok := gatewaypreauth.RequestModel(input.Req); ok {
			requestedModel = model
		}
	}
	sourceEndpointFamily := GatewayRequestEndpointFamily(input.Req)
	rawCandidateAccounts := input.RawCandidateAccounts
	if len(rawCandidateAccounts) == 0 && requestedModel != "" && input.LoadModelAwareCandidateAccounts != nil {
		loaded, err := input.LoadModelAwareCandidateAccounts(ctx, requestedModel, sourceEndpointFamily)
		if err != nil {
			return CandidateFilterOutput{}, err
		}
		if loaded != nil {
			rawCandidateAccounts = loaded
		}
	}
	if len(rawCandidateAccounts) == 0 {
		fallback, err := input.RouteCoordinator.RequestFallback(ctx, "no_candidate_accounts")
		if err != nil {
			return CandidateFilterOutput{}, err
		}
		if fallback.Attempted {
			return CandidateFilterOutput{Outcome: gatewaypreauth.CandidateOutcomeFallback, Reason: "no_candidate_accounts", Context: fallback.Context}, nil
		}
		if input.RecoverUnavailableCandidateAccounts != nil {
			recovered, err := input.RecoverUnavailableCandidateAccounts(ctx)
			if err != nil {
				return CandidateFilterOutput{}, err
			}
			if recovered != nil {
				rawCandidateAccounts = recovered
			}
		}
	}

	capabilityFilter := FilterGatewayAccountsByRequestCapability(
		input.Req, rawCandidateAccounts, p.engine.Driver,
		input.ClientStrategy.RequestClientCompatibility, "",
	)
	if capabilityFilter.SkippedCount > 0 {
		input.AuditCapture.AddGatewayMetadata("account_request_capability_filter", map[string]any{
			"skippedCount":               capabilityFilter.SkippedCount,
			"remainingCount":             len(capabilityFilter.Accounts),
			"reason":                     capabilityFilter.Reason,
			"requestClientCompatibility": input.ClientStrategy.RequestClientCompatibility,
		})
	}
	if len(rawCandidateAccounts) > 0 && len(capabilityFilter.Accounts) == 0 {
		reason := capabilityFilter.Reason
		if reason == "" {
			reason = "request_capability_mismatch"
		}
		fallback, err := input.RouteCoordinator.RequestFallback(ctx, reason)
		if err != nil {
			return CandidateFilterOutput{}, err
		}
		if fallback.Attempted {
			return CandidateFilterOutput{Outcome: gatewaypreauth.CandidateOutcomeFallback, Reason: reason, Context: fallback.Context}, nil
		}
		message := requestCapabilityMismatchMessage(reason)
		if err := input.RouteCoordinator.CompleteFailure(ctx, gatewayrouting.GatewayRouteFinalFailure{
			StatusCode: 503,
			Message:    message,
			ErrorType:  "service_unavailable",
			ErrorCode:  reason,
			ErrorPhase: "dispatch",
		}); err != nil {
			return CandidateFilterOutput{}, err
		}
		return CandidateFilterOutput{Outcome: gatewaypreauth.CandidateOutcomeCompleted}, nil
	}

	var modelFilter ModelFilterResult
	if input.BypassModelFilter {
		modelFilter = BypassGatewayModelFilter(capabilityFilter.Accounts, sourceEndpointFamily)
	} else {
		modelFilter = FilterGatewayAccountsByRequestedModel(capabilityFilter.Accounts, requestedModel, sourceEndpointFamily)
	}
	if shouldReloadModelAwareCandidates(requestedModel, modelFilter, input.LoadModelAwareCandidateAccounts) {
		modelAwareRawAccounts, err := input.LoadModelAwareCandidateAccounts(ctx, requestedModel, sourceEndpointFamily)
		if err != nil {
			return CandidateFilterOutput{}, err
		}
		if len(modelAwareRawAccounts) > 0 {
			modelAwareCapabilityFilter := FilterGatewayAccountsByRequestCapability(
				input.Req, modelAwareRawAccounts, p.engine.Driver,
				input.ClientStrategy.RequestClientCompatibility, "",
			)
			modelAwareModelFilter := FilterGatewayAccountsByRequestedModel(modelAwareCapabilityFilter.Accounts, requestedModel, sourceEndpointFamily)
			if modelAwareModelFilter.DirectMatchedCount > 0 ||
				modelAwareModelFilter.MappingMatchedCount > 0 ||
				(len(modelFilter.Accounts) == 0 && len(modelAwareModelFilter.Accounts) > 0) {
				capabilityFilter = modelAwareCapabilityFilter
				modelFilter = modelAwareModelFilter
				requestedModelAttribute := any(modelAwareModelFilter.RequestedModel)
				if requestedModelAttribute == "" {
					requestedModelAttribute = nil
				}
				familyAttribute := any(modelAwareModelFilter.SourceEndpointFamily)
				if familyAttribute == "" {
					familyAttribute = nil
				}
				reasonAttribute := any(modelAwareModelFilter.Reason)
				if reasonAttribute == "" {
					reasonAttribute = nil
				}
				input.AuditCapture.AddGatewayMetadata("account_model_candidate_window", map[string]any{
					"requestedModel":             requestedModelAttribute,
					"sourceEndpointFamily":       familyAttribute,
					"directMatchedCount":         modelAwareModelFilter.DirectMatchedCount,
					"mappingMatchedCount":        modelAwareModelFilter.MappingMatchedCount,
					"invalidModelConstraintCount": modelAwareModelFilter.InvalidModelConstraintCount,
					"remainingCount":             len(modelAwareModelFilter.Accounts),
				})
			}
		}
	}
	if modelFilter.SkippedCount > 0 || modelFilter.MappingMatchedCount > 0 {
		requestedModelAttribute := any(modelFilter.RequestedModel)
		if requestedModelAttribute == "" {
			requestedModelAttribute = nil
		}
		familyAttribute := any(modelFilter.SourceEndpointFamily)
		if familyAttribute == "" {
			familyAttribute = nil
		}
		reasonAttribute := any(modelFilter.Reason)
		if reasonAttribute == "" {
			reasonAttribute = nil
		}
		input.AuditCapture.AddGatewayMetadata("account_model_filter", map[string]any{
			"requestedModel":             requestedModelAttribute,
			"sourceEndpointFamily":       familyAttribute,
			"skippedCount":               modelFilter.SkippedCount,
			"limitedAccountCount":        modelFilter.LimitedAccountCount,
			"invalidModelConstraintCount": modelFilter.InvalidModelConstraintCount,
			"directMatchedCount":         modelFilter.DirectMatchedCount,
			"mappingMatchedCount":        modelFilter.MappingMatchedCount,
			"remainingCount":             len(modelFilter.Accounts),
			"reason":                     reasonAttribute,
		})
	}
	if len(capabilityFilter.Accounts) > 0 && len(modelFilter.Accounts) == 0 {
		reason := modelFilter.Reason
		if reason == "" {
			reason = "unsupported_model"
		}
		fallback, err := input.RouteCoordinator.RequestFallback(ctx, reason)
		if err != nil {
			return CandidateFilterOutput{}, err
		}
		if fallback.Attempted {
			return CandidateFilterOutput{Outcome: gatewaypreauth.CandidateOutcomeFallback, Reason: reason, Context: fallback.Context}, nil
		}
		message := GatewayModelFilterFailureMessage(modelFilter)
		if err := input.RouteCoordinator.CompleteFailure(ctx, gatewayrouting.GatewayRouteFinalFailure{
			StatusCode: 503,
			Message:    message,
			ErrorType:  "service_unavailable",
			ErrorCode:  reason,
			ErrorPhase: "dispatch",
		}); err != nil {
			return CandidateFilterOutput{}, err
		}
		return CandidateFilterOutput{Outcome: gatewaypreauth.CandidateOutcomeCompleted}, nil
	}

	return CandidateFilterOutput{
		Outcome:       gatewaypreauth.CandidateOutcomeAccounts,
		Accounts:      modelFilter.Accounts,
		ModelPriority: modelFilter.ModelPriority,
	}, nil
}

File: migration-backup-1/node/final-archive/backend/src/modules/gateway/dispatch/candidate-filter.ts
import type { Request, Response } from 'express'

import type { AuditCaptureContext } from '../audit/capture.service.js'
import type { OpenAIGatewayClientStrategyContext } from '../client-profiles/strategy.js'
import {
  filterGatewayAccountsByRequestCapability
} from './account-capability-filter.js'
import {
  filterGatewayAccountsByRequestedModel,
  gatewayAccountModelPriorityRank,
  type GatewayAccountModelPriority,
  type GatewayModelAccountFilterResult,
  gatewayModelFilterFailureMessage
} from './model-filter.js'
import type { UpstreamAccount } from '../protocols/openai-v1/route-helpers.js'
import { requestModel } from '../request/metadata.js'
import { gatewayRequestEndpointFamily } from '../protocols/openai-v1/model-mapping.js'
import type { GatewayFailureUsageContext } from '../usage/records.js'
import type { OpenAIGatewayDispatchContext } from '../request/preflight.js'
import type { GatewayRouteCoordinatorOwner } from '../routing/route-coordination.js'

export interface RequestCandidateFallbackResult {
  attempted: boolean
  context?: OpenAIGatewayDispatchContext
}

export type RequestCandidateFilterResult =
  | { outcome: 'accounts'; accounts: UpstreamAccount[]; modelPriority: GatewayAccountModelPriority }
  | { outcome: 'fallback'; reason: string; context?: OpenAIGatewayDispatchContext }
  | { outcome: 'completed' }

export async function filterOpenAIGatewayRequestCandidateAccounts(input: {
  req: Request
  res: Response
  auditCapture: AuditCaptureContext
  usageContext: GatewayFailureUsageContext
  startedAt: number
  rawCandidateAccounts: UpstreamAccount[]
  clientStrategy: OpenAIGatewayClientStrategyContext
  systemAccountId: string
  apiKeyId?: string
  groupId: string
  clientIp?: string
  endpoint: string
  bypassModelFilter?: boolean
  requestModelOverride?: string
  routeCoordinator: GatewayRouteCoordinatorOwner<OpenAIGatewayDispatchContext>
  recoverUnavailableCandidateAccounts?: () => Promise<UpstreamAccount[] | undefined>
  loadModelAwareCandidateAccounts?: (requestedModel: string, sourceEndpointFamily?: ReturnType<typeof gatewayRequestEndpointFamily>) => Promise<UpstreamAccount[] | undefined>
}): Promise<RequestCandidateFilterResult> {
  const requestedModel = input.requestModelOverride?.trim() || requestModel(input.req)
  const sourceEndpointFamily = gatewayRequestEndpointFamily(input.req)
  let rawCandidateAccounts = input.rawCandidateAccounts
  if (rawCandidateAccounts.length === 0 && requestedModel && input.loadModelAwareCandidateAccounts) {
    rawCandidateAccounts = await input.loadModelAwareCandidateAccounts(requestedModel, sourceEndpointFamily) ?? rawCandidateAccounts
  }
  if (rawCandidateAccounts.length === 0) {
    const fallback = await requestRouteFallback(input, 'no_candidate_accounts')
    if (fallback.attempted) {
      return { outcome: 'fallback', reason: 'no_candidate_accounts', context: fallback.context }
    }
    rawCandidateAccounts = await input.recoverUnavailableCandidateAccounts?.() ?? rawCandidateAccounts
  }

  let capabilityFilter = filterGatewayAccountsByRequestCapability(input.req, rawCandidateAccounts, {
    requestClientCompatibility: input.clientStrategy.requestClientCompatibility
  })
  if (capabilityFilter.skippedCount > 0) {
    input.auditCapture.addGatewayMetadata({
      label: 'account_request_capability_filter',
      metadata: {
        skippedCount: capabilityFilter.skippedCount,
        remainingCount: capabilityFilter.accounts.length,
        reason: capabilityFilter.reason,
        requestClientCompatibility: input.clientStrategy.requestClientCompatibility
      }
    })
  }
  if (rawCandidateAccounts.length > 0 && capabilityFilter.accounts.length === 0) {
    const reason = capabilityFilter.reason ?? 'request_capability_mismatch'
    const fallback = await requestRouteFallback(input, reason)
    if (fallback.attempted) {
      return { outcome: 'fallback', reason, context: fallback.context }
    }
    const message = requestCapabilityMismatchMessage(reason)
    await input.routeCoordinator.completeFailure({
      statusCode: 503,
      message,
      errorType: 'service_unavailable',
      errorCode: reason,
      errorPhase: 'dispatch'
    })
    return { outcome: 'completed' }
  }

  let modelFilter = input.bypassModelFilter
    ? bypassGatewayModelFilter(capabilityFilter.accounts, sourceEndpointFamily)
    : filterGatewayAccountsByRequestedModel(capabilityFilter.accounts, requestedModel, sourceEndpointFamily)
  if (shouldReloadModelAwareCandidates(requestedModel, modelFilter, input.loadModelAwareCandidateAccounts)) {
    const modelAwareRawAccounts = await input.loadModelAwareCandidateAccounts!(requestedModel!, sourceEndpointFamily)
    if (modelAwareRawAccounts?.length) {
      const modelAwareCapabilityFilter = filterGatewayAccountsByRequestCapability(input.req, modelAwareRawAccounts, {
        requestClientCompatibility: input.clientStrategy.requestClientCompatibility
      })
      const modelAwareModelFilter = filterGatewayAccountsByRequestedModel(modelAwareCapabilityFilter.accounts, requestedModel, sourceEndpointFamily)
      if (
        modelAwareModelFilter.directMatchedCount > 0
        || modelAwareModelFilter.mappingMatchedCount > 0
        || (modelFilter.accounts.length === 0 && modelAwareModelFilter.accounts.length > 0)
      ) {
        capabilityFilter = modelAwareCapabilityFilter
        modelFilter = modelAwareModelFilter
        input.auditCapture.addGatewayMetadata({
          label: 'account_model_candidate_window',
          metadata: {
            requestedModel: modelAwareModelFilter.requestedModel,
            sourceEndpointFamily: modelAwareModelFilter.sourceEndpointFamily,
            directMatchedCount: modelAwareModelFilter.directMatchedCount,
            mappingMatchedCount: modelAwareModelFilter.mappingMatchedCount,
            invalidModelConstraintCount: modelAwareModelFilter.invalidModelConstraintCount,
            remainingCount: modelAwareModelFilter.accounts.length
          }
        })
      }
    }
  }
  if (modelFilter.skippedCount > 0 || modelFilter.mappingMatchedCount > 0) {
    input.auditCapture.addGatewayMetadata({
      label: 'account_model_filter',
      metadata: {
        requestedModel: modelFilter.requestedModel,
        sourceEndpointFamily: modelFilter.sourceEndpointFamily,
        skippedCount: modelFilter.skippedCount,
        limitedAccountCount: modelFilter.limitedAccountCount,
        invalidModelConstraintCount: modelFilter.invalidModelConstraintCount,
        directMatchedCount: modelFilter.directMatchedCount,
        mappingMatchedCount: modelFilter.mappingMatchedCount,
        remainingCount: modelFilter.accounts.length,
        reason: modelFilter.reason
      }
    })
  }
  if (capabilityFilter.accounts.length > 0 && modelFilter.accounts.length === 0) {
    const fallback = await requestRouteFallback(input, modelFilter.reason ?? 'unsupported_model')
    if (fallback.attempted) {
      return { outcome: 'fallback', reason: modelFilter.reason ?? 'unsupported_model', context: fallback.context }
    }
    const message = gatewayModelFilterFailureMessage(modelFilter)
    await input.routeCoordinator.completeFailure({
      statusCode: 503,
      message,
      errorType: 'service_unavailable',
      errorCode: modelFilter.reason ?? 'unsupported_model',
      errorPhase: 'dispatch'
    })
    return { outcome: 'completed' }
  }

  return {
    outcome: 'accounts',
    accounts: modelFilter.accounts,
    modelPriority: modelFilter.modelPriority
  }
}

async function requestRouteFallback(
  input: {
    routeCoordinator: GatewayRouteCoordinatorOwner<OpenAIGatewayDispatchContext>
  },
  reason: string
): Promise<RequestCandidateFallbackResult> {
  return input.routeCoordinator.requestFallback(reason)
}

function bypassGatewayModelFilter(
  accounts: UpstreamAccount[],
  sourceEndpointFamily: ReturnType<typeof gatewayRequestEndpointFamily>
): GatewayModelAccountFilterResult {
  return {
    accounts,
    skippedCount: 0,
    limitedAccountCount: accounts.filter((account) => (account.supportedModels?.length ?? 0) > 0).length,
    invalidModelConstraintCount: 0,
    directMatchedCount: accounts.length,
    mappingMatchedCount: 0,
    sourceEndpointFamily,
    modelPriority: {
      sourceEndpointFamily,
      rankByAccountId: new Map(accounts.map((account) => [account.id, gatewayAccountModelPriorityRank.direct]))
    }
  }
}

function shouldReloadModelAwareCandidates(
  requestedModel: string | undefined,
  modelFilter: {
    directMatchedCount: number
    mappingMatchedCount: number
  },
  loader: ((requestedModel: string, sourceEndpointFamily?: ReturnType<typeof gatewayRequestEndpointFamily>) => Promise<UpstreamAccount[] | undefined>) | undefined
): boolean {
  return Boolean(
    requestedModel
    && loader
    && modelFilter.directMatchedCount === 0
    && modelFilter.mappingMatchedCount === 0
  )
}

function requestCapabilityMismatchMessage(reason: string): string {
  if (reason === 'anthropic_native_group_openai_compatible_request') {
    return '当前 API Key 绑定的是 Anthropic 原生分组，不兼容 Codex / OpenAI 请求路径；请改用 Anthropic /v1/messages 客户端，或绑定支持 OpenAI Responses / Chat Completions 的分组'
  }
  return '当前分组无账户支持请求路径或客户端协议'
}
